sktree/tree/_oblique_splitter.pyx

```diff
@@ -453,3 +477,198 @@ cdef class BestObliqueSplitter(ObliqueSplitter):
+
+        # XXX: Just to initialize stuff
+        # self.feature_weights = np.ones((self.n_features,), dtype=DTYPE_t) / self.n_features
+        return 0
+
+    # overwrite the node_soplit method with random threshold selection
+    cdef int node_split(
+        self,
+        double impurity,
+        SplitRecord* split,
+        SIZE_t* n_constant_features,
+    ) except -1 nogil:
+        """Find the best random split on node samples[start:end]
+
+        Returns -1 in case of failure to allocate memory (and raise MemoryError)
+        or 0 otherwise.
+        """
+        # typecast the pointer to an ObliqueSplitRecord
+        cdef ObliqueSplitRecord* oblique_split = <ObliqueSplitRecord*>(split)
+
+        # Draw random splits and pick the best_split
+        cdef SIZE_t[::1] samples = self.samples
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+
+        # pointer array to store feature values to split on
+        cdef DTYPE_t[::1]  feature_values = self.feature_values
+        cdef SIZE_t max_features = self.max_features
+        cdef SIZE_t min_samples_leaf = self.min_samples_leaf
+        cdef double min_weight_leaf = self.min_weight_leaf
+
+        # keep track of split record for current_split node and the best_split split
+        # found among the sampled projection vectors
+        cdef ObliqueSplitRecord best_split, current_split
+        # cdef double current_proxy_improvement = -INFINITY
+        # cdef double best_proxy_improvement = -INFINITY
+
+        cdef SIZE_t feat_i, p      # index over computed features and start/end
+        cdef SIZE_t partition_end
+        cdef DTYPE_t temp_d, threshold         # to compute a projection feature value
+
+        # instantiate the split records
+        _init_split(&best_split, end)
+
+        # Sample the projection matrix
+        self.sample_proj_mat(self.proj_mat_weights, self.proj_mat_indices)
+
+        # For every vector in the projection matrix
+        for feat_i in range(max_features):
+            # Projection vector has no nonzeros
+            if self.proj_mat_weights[feat_i].empty():
+                continue
+
+            # XXX: 'feature' is not actually used in oblique split records
+            # Just indicates which split was sampled
+            current_split.feature = feat_i
+            current_split.proj_vec_weights = &self.proj_mat_weights[feat_i]
+            current_split.proj_vec_indices = &self.proj_mat_indices[feat_i]
+
+            # Compute linear combination of features and then
+            # sort samples according to the feature values.
+            self.compute_features_over_samples(
+                start,
+                end,
+                samples,
+                feature_values,
+                &self.proj_mat_weights[feat_i],
+                &self.proj_mat_indices[feat_i]
+            )
+
+            # Sort the samples
+            sort(&feature_values[start], &samples[start], end - start)
```

Comment:
See here you are still sorting.

Reply:
# Draw a random threshold
        current_split.threshold = rand_uniform(
            min_feature_value,
            max_feature_value,
            random_state,
        )

In `node_split_random` in `Splitter`, it samples a random (uniform) threshold between `min_feature_value` and `max_feature_value` as above and find and set `.pos` according to it by `partitioner. partition_samples. To find `min_feature_value` and `max_feature_value`, it uses `partitioner.find_min_max`, which runs iteratively each feature value and is basically the same as sorting.  I was going to use that to get `min_feature_value` and `max_feature_value` then follow the same logic as `node_split_random`. But I had a question: can we just randomly sample `current_split.pos` from `rand_int(start+min_samples_leaf, end, &self.rand_r_state)`? Is that a bad sampling strategy?  Is either better than the other?  If the latter is an acceptable sampling method, it'll be faster, (and no sorting).

Reply:
I'm not sure I understand the question. You wouldn't randomly sample `pos`. Also finding min/max is not the same as sorting. Sorting is O(nlog(n)). Finding min/max is O(n). 

The pseudocode for RandomSplitter.node_split is the following. The only difference is that `compute_features()` computes a projection of the data, rather than just choosing a column of the data. I think sklearn's random splitter does some extra stuff like keeping track of what columns contain constant values, whereas we don't do that.

```
for idx in range(max_features):
    # for each projection, compute the features over all samples
    # (n_samples,) vector
    current_feature_values = compute_features_over_samples

    # find min and max of current feature values
    min, max = find_min_max(current_feature_values)

    # draw random thresohld
    current_split.threshodl = rand_uniform(min, max, randstate)

    # partition samples given the threshold by moving feature values and samples inplace
    current_split.pos = partition_samples()

    # evaluate split
    criterion.reset()
    criterion.update(current_split.pos)

    current_proxy_improvement = self.criterion.proxy_impurity_improvement()
    if current_proxy_improvement > best_proxy_improvement:
        best_proxy_improvement = current_proxy_improvement
        best = current  # copy
    
    # reorganize samples inplace samples[start:best.pos] + samples[best.pos:end]
    # copy of the previous code
```

Possibly some good papers on this topic are the extra trees publication linked in the GH issue and the original random forest paper by Breiman.